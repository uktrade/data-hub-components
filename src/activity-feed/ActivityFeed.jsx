import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { SPACING } from '@govuk-react/constants'

import Activity from './Activity'
import ActivityFeedHeader from './ActivityFeedHeader'
import ActivityFeedFilters from './ActivityFeedFilters'
import ActivityFeedShowAll from './ActivityFeedShowAll'
import ActivityFeedPagination from './ActivityFeedPagination'

const ActivityFeedContainer = styled('div')`
  margin: ${SPACING.SCALE_2} 0;
`

const ActivityFeedCardList = styled('ol')`
  list-style-type: none;
  padding: 0;
  margin-top: ${SPACING.SCALE_2};

  & > li {
    margin-bottom: ${SPACING.SCALE_2};
  }
`

export default class ActivityFeed extends React.Component {
  static propTypes = {
    onLoadMore: PropTypes.func,
    sendQueryParams: PropTypes.func,
    children: PropTypes.node,
    activities: PropTypes.arrayOf(PropTypes.object),
    activityTypeFilters: PropTypes.object,
    hasMore: PropTypes.bool,
    isLoading: PropTypes.bool,
    actions: PropTypes.node,
    totalActivities: PropTypes.number,
    isGlobalUltimate: PropTypes.bool,
    dnbHierarchyCount: PropTypes.number,
    isTypeFilterFlagEnabled: PropTypes.bool,
    isGlobalUltimateFlagEnabled: PropTypes.bool,
  }

  static defaultProps = {
    onLoadMore: () => {},
    sendQueryParams: () => {},
    children: null,
    activities: [],
    activityTypeFilters: {},
    hasMore: false,
    isLoading: false,
    actions: null,
    totalActivities: 0,
    isGlobalUltimate: false,
    dnbHierarchyCount: null,
    isTypeFilterFlagEnabled: false,
    isGlobalUltimateFlagEnabled: false,
  }

  constructor(props) {
    const {
      allActivity,
      myActivity,
      externalActivity,
      dataHubActivity,
    } = props.activityTypeFilters

    super(props)

    this.state = {
      activityTypeFilter: dataHubActivity ? dataHubActivity.value : '',
      showDnbHierarchy: false,
      showDetails: false,
      activityTypeFilters: [
        allActivity,
        myActivity,
        externalActivity,
        dataHubActivity,
      ],
    }

    this.onActivityTypeFilterChange = this.onActivityTypeFilterChange.bind(this)
    this.onShowDetailsClick = this.onShowDetailsClick.bind(this)
    this.showActivitiesFromAllCompanies = this.showActivitiesFromAllCompanies.bind(
      this
    )
  }

  onActivityTypeFilterChange(e) {
    const activityTypeFilter = e.target.value
    const { sendQueryParams } = this.props
    const { showDnbHierarchy } = this.state

    sendQueryParams({
      activityTypeFilter,
      showDnbHierarchy,
    })

    this.setState({
      activityTypeFilter,
    })
  }

  showActivitiesFromAllCompanies(e) {
    const showDnbHierarchy = e.target.checked
    const { sendQueryParams } = this.props
    const { activityTypeFilter } = this.state

    sendQueryParams({
      activityTypeFilter,
      showDnbHierarchy,
    })

    this.setState({
      showDnbHierarchy,
    })
  }

  onShowDetailsClick(e) {
    this.setState({
      showDetails: e.target.checked,
    })
  }

  render() {
    const {
      activities,
      onLoadMore,
      hasMore,
      isLoading,
      actions,
      children,
      totalActivities,
      isGlobalUltimate,
      dnbHierarchyCount,
      isTypeFilterFlagEnabled,
      isGlobalUltimateFlagEnabled,
    } = this.props

    const {
      activityTypeFilters,
      activityTypeFilter,
      showDetails,
      showDnbHierarchy,
    } = this.state

    const hasFilters = isTypeFilterFlagEnabled || isGlobalUltimateFlagEnabled

    return (
      <ActivityFeedContainer>
        <ActivityFeedHeader
          totalActivities={totalActivities}
          actions={actions}
        />

        {hasFilters && (
          <ActivityFeedFilters
            activityTypeFilters={activityTypeFilters}
            activityTypeFilter={activityTypeFilter}
            onActivityTypeFilterChange={this.onActivityTypeFilterChange}
            showActivitiesFromAllCompanies={this.showActivitiesFromAllCompanies}
            isGlobalUltimate={isGlobalUltimate}
            dnbHierarchyCount={dnbHierarchyCount}
            isTypeFilterFlagEnabled={isTypeFilterFlagEnabled}
            isGlobalUltimateFlagEnabled={isGlobalUltimateFlagEnabled}
          />
        )}

        {activities.length > 0 && (
          <ActivityFeedShowAll
            onChange={this.onShowDetailsClick}
            checked={showDetails}
          >
            Show details for all activities
          </ActivityFeedShowAll>
        )}

        <ActivityFeedCardList>
          {activities.map((activity) => (
            <li key={activity.id}>
              <Activity
                activity={activity}
                showDetails={showDetails}
                showDnbHierarchy={showDnbHierarchy}
              />
            </li>
          ))}
        </ActivityFeedCardList>

        {hasMore && (
          <ActivityFeedPagination
            isLoading={isLoading}
            onLoadMore={onLoadMore}
          />
        )}

        {children}
      </ActivityFeedContainer>
    )
  }
}
