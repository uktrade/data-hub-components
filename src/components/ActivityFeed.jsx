import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { SPACING } from '@govuk-react/constants'

import ActivityFeedCard from './ActivityFeedCard'
import ActivityFeedHeader from './ActivityFeedHeader'
import ActivityFeedPagination from './ActivityFeedPagination'

const ActivityFeedContainer = styled('div')`
  margin: ${SPACING.SCALE_2};
`

const ActivityFeedCardList = styled('ol')`
  list-style-type: none;
  padding: 0;
  
  & > li {
    margin-bottom: ${SPACING.SCALE_2};
  }
`

export default class ActivityFeed extends React.Component {
  static propTypes = {
    activities: PropTypes.arrayOf(PropTypes.object).isRequired,
    onLoadMore: PropTypes.func,
  }

  render() {
    const { activities, onLoadMore, isLoading } = this.props

    return (
      <ActivityFeedContainer>
        <ActivityFeedHeader totalCards={activities.length} />

        <ActivityFeedCardList>
          {activities.map(activity => {
            return <li key={activity.id}>
              <ActivityFeedCard activity={activity} />
            </li>
          })}
        </ActivityFeedCardList>

        {onLoadMore && <ActivityFeedPagination isLoading={isLoading} onLoadMore={onLoadMore}/>}
      </ActivityFeedContainer>
    )
  }
}
