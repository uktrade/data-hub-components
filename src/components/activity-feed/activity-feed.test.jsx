import React from 'react'
import renderer from 'react-test-renderer'

import ActivityFeed from './activity-feed'

test('renders empty feed', () => {
  const tree = renderer
    .create(<ActivityFeed activities={[]} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
