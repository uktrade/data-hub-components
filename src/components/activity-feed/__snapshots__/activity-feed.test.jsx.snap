// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty feed 1`] = `
Array [
  <article>
    0
     activities
    <button>
      Add activity
    </button>
    <form>
      <div>
        Show me 
        <label>
          <input
            type="checkbox"
          />
          My filters
        </label>
      </div>
      <div>
        Activity type 
        <select>
          <option>
            All activity
          </option>
        </select>
      </div>
    </form>
  </article>,
  <button>
    Show more activity
  </button>,
]
`;
